### PR TITLE
fix(mobile): auto-connect when the bridge comes online while the app waits

### PR DESCRIPTION
**Pull request by daniil-shumko**
## Problem

Reported bug: start the app with the bridge **offline** → it shows the "Set up Sesori Bridge" / bridge-disconnected screen. Start the bridge → **the app does not connect; it requires a pull-to-refresh.**

By contrast, a bridge that drops *after* a live connection auto-recovers instantly. The two cases looked identical in the UI but used different connection states:

| | launched while bridge offline | bridge dropped after connecting |
|---|---|---|
| state | `ConnectionDisconnected` | `ConnectionBridgeOffline` |
| relay socket | none (torn down) | kept open |
| bridge-status watcher | never armed | armed |
| auto-connect on bridge-online | ❌ no | ✅ yes |

## Root cause (client-side)

The **relay already** keeps a phone socket open when no bridge is present and pushes `bridge_connected` the moment one appears (verified in `sesori_relay_server` `handler.go` — `handlePhone` parks the phone and `handleBridge` fans out `bridge_connected`). The gap was that the client tore its **own** socket down: during the initial handshake the relay's `bridge_disconnected` frame errored the key-exchange completer, which threw out of `RelayClient.connect()` into the teardown path — skipping the block that retains the `RelayClient` and arms `_subscribeBridgeStatus`. With no socket and no watcher, the later `bridge_connected` had nowhere to land.

## Fix

Treat "relay + auth socket established, but no bridge yet" as a first-class, socket-alive outcome instead of a fatal error — then reuse the existing, already-tested `bridge_connected → _reconnectRelayWithRefresh` recovery path.

- **`relay_client.dart`**: `connect()` returns a `RelayConnectOutcome`. On the initial `bridge_disconnected` it resolves to `bridgeAbsent` (keeping the socket open and the stored room key) instead of tearing down. A new `bridgeOffline` connection state keeps `isConnected == false`, so `RelayHttpApiClient` cleanly returns a "relay not connected" response rather than routing through the half-open client. Added `onSocketClosed` so a drop while parked (no SSE stream to surface it) is still detected.
- **`connection_service.dart`**: commits the `bridgeAbsent` outcome as `ConnectionBridgeOffline` with the bridge-status watcher armed (skipping the health probe and SSE, which need the room key). `_onAppResumed` now reconnects a stale parked socket on foreground; a still-absent bridge simply re-parks.

**No relay or auth-server change is required** — the relay already implements the "hold the phone, pair on arrival" behavior.

## Security

Unchanged. The relay still only forwards opaque encrypted frames (it never participates in the DH). On `bridge_connected` the phone runs a **fresh ephemeral DH** — no buffered/replayed handshake material — so forward secrecy is preserved.

## Testing

- `flutter analyze`: clean. `flutter test` (module_core): **403 passing**, including 3 new tests — park in `ConnectionBridgeOffline` (no health probe / SSE), bridge-online → reconnect to `ConnectionConnected`, and socket-drop-while-parked recovery.
- **Live-verified on the iPhone 17 Pro Max simulator**: launch with bridge offline → parked in bridge-offline → start `sesori-bridge` → app auto-connects to the projects list with **no pull-to-refresh** and no user interaction.

```
Relay connected but bridge is offline — holding socket open   ← fix: socket kept open
[ProjectList] connection status: ConnectionBridgeOffline
Relay: bridge came online                                     ← relay push
Bridge came back online — reconnecting to re-establish encryption
Relay key exchange complete, room key persisted               ← fresh DH
[ProjectList] connection status: ConnectionConnected          ← auto-connected
```

## Follow-ups (out of scope, separate PRs)

- Relay token-expiry deadline-close (`4007`) so long-parked sockets re-validate (`authResult.Expiry` is currently dead code in the relay).
- Client resilience: jittered backoff + bridge-flap debounce.

---
## Summary by cubic
Fixes the mobile app not auto-connecting when launched while the bridge is offline. The app now keeps the relay socket open and auto-connects as soon as the bridge comes online—no pull-to-refresh.

- **Bug Fixes**
  - `RelayClient.connect()` no longer fails when the bridge is absent; it keeps the socket up with no session (`isConnected == false`), so API calls don’t route through a half-open client.
  - `ConnectionService`: when `connect()` returns with `isConnected == false` and the transport state is `connected`, commit `ConnectionBridgeOffline`, arm the bridge-status watcher, skip health probe/SSE, and watch `onSocketClosed` to recover drops while parked.
  - Reconnect immediately on `bridge_connected` (skip backoff delay).
  - On app resume, proactively reconnect likely-stale parked sockets; if the bridge is still absent, it re-parks.
  - Prevent uncaught async errors by clearing the unused handshake completer; tests cover park-without-health/SSE, bridge-online reconnect, socket-close recovery, and stub `RelayClient.connectionState` for the park guard.

<sup>Written for commit bb2626f596ad4b97ef38a7783d5ea1e6e826bc20. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/sesori-ai/sesori_apps_monorepo/pull/296?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: mobile/module_core/lib/src/capabilities/relay/relay_client.dart [hunk 1]
```diff
@@ -9,11 +9,39 @@ import "package:web_socket_channel/web_socket_channel.dart";
...
+  /// Socket established AND the E2E key exchange (or resume) with the bridge
+  /// completed — the client is ready for encrypted requests.
+  connected,
+
+  /// Socket established and authenticated, but no bridge is currently in the
+  /// account group, so the E2E key exchange could not complete. The socket is
+  /// kept open; the relay pushes a bridge_connected control frame when a bridge
+  /// appears, at which point a fresh handshake can run.
+  bridgeAbsent,
+}
```
Comment (AlexDochioiu) [inline mobile/module_core/lib/src/capabilities/relay/relay_client.dart]: This smells. I feel like it should simply be

```
enum RelayConnectionState {
  disconnected,
  disconnecting,
  connecting,
  connected;
};

enum BridgeStatus { online, offline }
```
Note: Not even sure if we need disconnecting for relay state. Is this ever used?

Code looks sloppy and overcomplicated + the 3 enums make no logical sense.

We must not conflate any relay state with bridge state. Not sure if relevant here, but I'll also state it clearly to make sure this does not apply: we must also NOT conflate bridge status with plugin status. Bridge can be online even if plugin is not available/etc.

File: mobile/module_core/lib/src/capabilities/relay/relay_client.dart [hunk 2]
```diff
@@ -539,9 +619,10 @@ class RelayClient {
           logd("Relay: bridge went offline");
           final firstBinaryMessage = _firstBinaryMessage;
           if (firstBinaryMessage != null && !firstBinaryMessage.isCompleted) {
-            firstBinaryMessage.completeError(
-              StateError("Bridge is offline - cannot complete key exchange"),
-            );
+            // Bridge absent during the handshake. Signal a distinct, non-fatal
+            // outcome so connect() keeps the socket open and waits for a later
+            // bridge_connected instead of tearing the socket down.
+            firstBinaryMessage.completeError(const _BridgeOfflineDuringHandshake());
```
Comment (AlexDochioiu) [inline mobile/module_core/lib/src/capabilities/relay/relay_client.dart:610]: Solution smells. I'd rather get an explicit instant `bridge_disconnected` from the server if we connect and no bridge is available. If needed, make those changes and raise a separate PR on the bridge repo.

P.s. I think as soon as a client connected, relay server should always emit the current bridge state to the client (bridge connected / bridge disconnected)

File: mobile/module_core/lib/src/capabilities/relay/relay_client.dart [hunk 3]
```diff
@@ -133,11 +171,27 @@ class RelayClient {
...
+      // Keep the socket alive (do NOT tear it down) and report bridge-absent so
+      // ConnectionService can park in ConnectionBridgeOffline and let the relay's
+      // bridge_connected control frame drive a reconnect when a bridge appears.
+      if (_disposed) return RelayConnectOutcome.bridgeAbsent;
+      // Socket + auth are up but there is no E2E session yet, so isConnected
+      // stays false — RelayHttpApiClient won't route requests through this
+      // half-open client until a bridge appears and the key exchange completes.
+      _connectionState = RelayClientConnectionState.bridgeOffline;
+      // Re-arm the binary completer so a later handshake can run on this socket.
+      _firstBinaryMessage = Completer<Uint8List>();
```
[the remark this answers is not in the document]
Comment (daniil-shumko) [reply mobile/module_core/lib/src/capabilities/relay/relay_client.dart]: Fixed in dd80056 — set `_firstBinaryMessage = null` in the bridge-absent branch instead of re-arming it, exactly as suggested. Confirmed the unobserved-completer path (a bridge flap or socket close while parked would `completeError` a future nobody awaits → uncaught zone error). The `!= null` guards in `_onTextFrame`/`_onSocketDone` now no-op while parked, and recovery builds a fresh `RelayClient` so the completer is never reused. Thanks for the catch.

File: mobile/module_core/lib/src/capabilities/relay/relay_client.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (daniil-shumko) [reply mobile/module_core/lib/src/capabilities/relay/relay_client.dart:610]: Confirmed and fixed in dd80056. The parked re-armed completer is gone — set to `null` instead — so a bridge flap before the reconnect can no longer error an unobserved future. Good catch.

File: mobile/module_core/lib/src/capabilities/relay/relay_client.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (daniil-shumko) [reply mobile/module_core/lib/src/capabilities/relay/relay_client.dart]: Valid (P1) — fixed in dd80056 by setting `_firstBinaryMessage = null` in the bridge-absent branch rather than re-arming. Recovery always builds a fresh `RelayClient`, so it is never reused; the `_onTextFrame` guard now safely no-ops while parked.

File: mobile/module_core/lib/src/capabilities/relay/relay_client.dart (hunk 1 above)
Comment (daniil-shumko) [reply mobile/module_core/lib/src/capabilities/relay/relay_client.dart]: Agreed — fixed in dd80056. Removed the conflation: `RelayClientConnectionState` is back to the four transport states (dropped `bridgeOffline`), and I also dropped the `RelayConnectOutcome` enum. `isConnected` now means "socket up AND an E2E session exists" (`connected && _sessionEncryptor != null`), so the bridge-absent park reports `false` (and `RelayHttpApiClient` won't route through it) without any bridge concept leaking into the transport state. `connect()` returns `void`; the caller inspects `isConnected`.

(Re `disconnecting`: it's pre-existing and still used by `disconnect()`, so I left it to keep this PR scoped — happy to revisit separately. And agreed on not conflating bridge status with plugin status; this change keeps relay/bridge/session concerns separate.)

File: mobile/module_core/lib/src/capabilities/relay/relay_client.dart (hunk 2 above)
Comment (daniil-shumko) [reply mobile/module_core/lib/src/capabilities/relay/relay_client.dart:610]: The relay already does exactly that: when a phone joins with no bridge in the account group it sends an explicit, instant `bridge_disconnected` (sesori_relay_server `internal/relay/handler.go` → `handlePhone`), and it pushes `bridge_connected` to waiting phones the moment a bridge arrives. So no bridge-repo change is needed — the only bug was the client treating that explicit frame as a fatal handshake error and tearing the socket down. It now consumes it to park (socket kept open) and reconnects on the later `bridge_connected` (now `immediate: true`, so no backoff wait).

On the tear-down-then-reconnect being inelegant — agreed. The cleaner in-place re-key (reuse the open socket, run a fresh DH on it, no new WSS / token refresh) is filed as a fast-follow: an independent review flagged real regression risk to the shared post-connect-drop recovery path, so I kept this PR to the lower-risk reuse of the already-tested reconnect path. Happy to fold the in-place version into this PR instead if you'd prefer.

**Comment by daniil-shumko** [comment]
Pushed `dd80056` addressing review:

- **Uncaught async error (Gemini/Codex/cubic):** the parked branch no longer re-arms `_firstBinaryMessage` — it's set to `null`, so a bridge flap or socket-close while parked can't `completeError` an unobserved future.
- **No relay/bridge state conflation (@AlexDochioiu):** dropped `RelayClientConnectionState.bridgeOffline` and the `RelayConnectOutcome` enum. Transport enum is back to four states; `isConnected` now derives from the session (`connected && _sessionEncryptor != null`), so the bridge-absent park reports `false` and `RelayHttpApiClient` won't route through it. `connect()` returns `void`; the caller inspects `isConnected`.
- **Instant reconnect:** the `bridge_connected` handler now uses `immediate: true`, removing the ~1s backoff dead time on a provably-present bridge.

Also: the relay already sends an explicit instant `bridge_disconnected` on a bridge-less join, so no bridge-repo change is needed.

Re-verified live on iPhone 17 Pro Max (cold-start offline → park → start bridge → immediate auto-connect, no pull-to-refresh, no uncaught errors). `flutter analyze` clean; all 403 module_core tests pass.

Fast-follows (separate PRs): in-place re-key on the parked socket (removes the tear-down+reconnect entirely); `immediate`/jitter on the post-connect drop path; thundering-herd jitter; relay-side JWT re-validation on live sockets.

File: mobile/module_core/lib/src/capabilities/relay/relay_client.dart (hunk 2 above)
Comment (AlexDochioiu) [reply mobile/module_core/lib/src/capabilities/relay/relay_client.dart:610]: Let's do this as a new follow up PR.

File: mobile/module_core/lib/src/capabilities/server_connection/connection_service.dart [hunk 4]
```diff
@@ -275,6 +276,37 @@ class ConnectionService {
...
+      // connect() returned but isConnected is false: the relay accepted and is
+      // holding our socket open, but no bridge is in the account group yet, so
+      // no E2E session exists. Commit the socket and arm the bridge-status
+      // watcher, skipping the health probe (no session encryptor) and SSE (needs
+      // the room key). When a bridge appears the relay pushes bridge_connected;
+      // _subscribeBridgeStatus then drives a reconnect that runs a fresh key
+      // exchange — the same recovery path as a bridge that drops after a live
+      // connection. No await runs between the staleness check above and this
+      // commit, so there is no superseding race to re-check.
+      if (!relayClient.isConnected) {
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply mobile/module_core/lib/src/capabilities/server_connection/connection_service.dart]: [Sesori reply] Addressed (in commit dd02c9d7)

Added a transport-state guard: the park branch now requires `connectionState == RelayClientConnectionState.connected && !isConnected`. The bridge-absent park sets the state to connected with no session encryptor, so it still parks correctly; the only newly-excluded path is a disposed/early return where the state stays `connecting` — which should not park. Note the bot's primary scenario (a just-dropped socket returning from connect()) does not actually occur: a socket close mid-handshake completeErrors the handshake completer, which throws and is handled by the outer catch/rethrow rather than returning normally. The guard is kept as a low-cost hardening for the disposed edge case.

File: mobile/module_core/test/capabilities/server_connection/connection_service_reconnect_test.dart [hunk 5]
```diff
@@ -630,5 +633,156 @@ void main() {
       verify(reconnectClient.disconnect).called(greaterThanOrEqualTo(1));
       verifyNever(() => reconnectClient.sendRequest(any()));
     });
+
+    test("connect with bridge absent parks in ConnectionBridgeOffline without health probe or SSE", () async {
+      final relayClient = MockRelayClient();
+      when(relayClient.connect).thenAnswer((_) async {});
+      // Bridge absent: connect() returns but isConnected stays false.
+      when(() => relayClient.isConnected).thenReturn(false);
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply mobile/module_core/test/capabilities/server_connection/connection_service_reconnect_test.dart:641]: [Sesori reply] Addressed (in commit bb2626f5)

Correct diagnosis — this was the cause of the CI failure on the previous commit. `connectionState` is now stubbed as `RelayClientConnectionState.connected` on every MockRelayClient that returns from connect(): the bridge-absent/parked and reparked mocks (which keep isConnected false) as well as the already-connected mocks, across connection_service_reconnect_test.dart, connection_service_auth_state_test.dart, and connection_service_stale_test.dart. Full module_core suite (403 tests) passes locally.
